File: reuse_java/src/main/java/com/kmmania/Jumper.java
package com.kmmania;

/**
 * The Jumper class implements the {@link AthleteInterface} by composition and represents an athlete
 * specializing in jumping.
 *
 * @author Dac Linh Giao NGUYEN
 * @version 0.1
 * @see AthleteInterface
 * @see AthleteImpl
 */
public class Jumper {
    private final String firstName;
    private final String lastName;
    private final AthleteInterface athlete;

    /**
     * Constructs a Jumper object with the specified first name, last name, and specialty.
     *
     * @param firstName the first name of the jumper
     * @param lastName  the last name of the jumper
     * @param athlete   An instance of the {@link AthleteInterface} representing the distance runner.
     */
    public Jumper(String firstName, String lastName, AthleteInterface athlete) {
        this.firstName = firstName;
        this.lastName = lastName;
        this.athlete = athlete;
    }

    public String getFirstName() {
        return firstName;
    }

    public String getLastName() {
        return lastName;
    }

    /**
     * Displays the jumper's information including name and specialty.
     */
    public void display() {
        System.out.printf("My name is %s %s. %s\n",
                this.firstName,
                this.lastName,
                this.athlete.mySpeciality()
        );
    }
}

File: reuse_python/src/jumper.py
"""
Athlete Reusability Project

This project demonstrates different ways to reuse classes in Python:
- Inheritance
- Composition
- Composition through interface implementation

Classes:
    Athlete: Represents an athlete with a specific specialty.
    Sprinter: Represents a sprinter, inheriting from Athlete.
    Thrower: Represents a thrower, utilizing Athlete class through composition.
    AthleteInterface: Interface defining the method for athlete specialty.
    AthleteBase: Class implementing the AthleteInterface.
    DistanceRunner: Represents a distance runner, using Athlete interface through composition.
    Jumper: Represents a jumper, using Athlete interface through composition.

Usage:
    Import the required classes and create instances of different types of athletes.
    Each athlete instance can display information about themselves using the `display()` method.


Author: Dac Linh Giao NGUYEN
Date: 2024

"""


class Jumper:
    """Class representing a jumper."""

    def __init__(self, first_name, last_name, athlete):
        """
        Initializes a Jumper object with a given first name, last name, and athlete instance.

        Args:
            first_name (str): The first name of the jumper.
            last_name (str): The last name of the jumper.
            athlete (AthleteBase): An Athlete object representing the jumper's specialty.
        """
        self.first_name = first_name
        self.last_name = last_name
        self.athlete = athlete

    def display(self):
        """
        Displays information about the jumper.

        Returns:
            None
        """
        print(f"My name is {self.first_name} {self.last_name}. {self.athlete.my_speciality()}")
